[MP] MyDiagnoses-Screen

MPucheta opened the pull request:
- Implemented 'mis solicitudes' screen as it shows in the wireframe

- I added a cut to the text (dots at the end), as too much description would result in the cards beings too big and contradicting what the wireframe intends to show.

- I did not include the 'Abierto' and such indications. It seems more like the responsibility of the 'diagnose details' screen (I think we would need to use the cache too).

Screencasts:

![image](https://user-images.githubusercontent.com/35865469/71300245-ad0fec80-2371-11ea-81dd-c771130234df.png)
![solicitudes](https://user-images.githubusercontent.com/35865469/71300334-65d62b80-2372-11ea-8fec-a8f1f02206da.gif)

MPucheta committed:
[MP/MyDiagnoses-Screen][ADDED] added 'my solicitudes' implementation according to wireframe specification

MPucheta committed:
[MP/MyDiagnoses-Screen][ADDED] added 'DiagnoseRequest' as default route of MainApp

MPucheta commented on src/screens/MainApp/MyDiagnoses/index.js:
```diff
@@ -0,0 +1,81 @@
+    // Tab navigator doesn't unmount components, we need this to refetch need data when pressing this tab.
+    // alternative is to add a query snapshot listener and a MobX store
+    toggleRefetch(!refetchTrigger)
+  }
```
I really do not like this. I believe it's ok for now and it really doesn't incur in more database hits (there is a window of 30 minutes of free hits if the data did not change, I could not find this in the docs though, so I ping @nahudalla for insight) But it seems hacky to force a re-render this way. I think the best way to do this is with mobx and introducing reactiveness, but it's a little expensive on time.

nahudalla replied:
Maybe this will solve the problem: https://reactnavigation.org/docs/en/next/use-focus-effect.html

_We're using React Navigation, right?_

MPucheta replied:
I will give it a shot

nahudalla commented on src/mixins/diagnose/index.js:
```diff
@@ -0,0 +1,7 @@
+
+export const sortDiagnosesByMostRecentUpdate = (diagnoses) => (
+  diagnoses.sort((a, b) => b.updatedAt.toMillis() - a.updatedAt.toMillis())
+)
```
What do you think about generalizing the sorting this way?

```javascript
const sortAscending = (array, getSortValue = a => a) => (
  array.sort((a, b) => getSortValue(a) - getSortValue(b))
)

export const sortDiagnosesByMostRecentCreation = (diagnoses) => (
  sortAscending(diagnoses, a => a.createdAt.toMillis())
)

export const sortDiagnosesByMostRecentUpdate = (diagnoses) => (
  sortAscending(diagnoses, a => a.updatedAt.toMillis())
)
```

MPucheta replied:
It's quite good. There may be some cases that unexpected behavior occur if the `getSortValue` does not return something that makes sense with the `-` operator, but it's up to the developer to take care of this.

I will implement this right away

nahudalla commented on src/screens/MainApp/MyDiagnoses/index.js:
```diff
@@ -0,0 +1,81 @@
+    } catch (error) {
+      setDiagnoses(null)
+    }
+    setDownloadingDiagnoses(false)
```
Please move this line to the `finally` block

nahudalla commented on src/screens/MainApp/MyDiagnoses/index.js:
```diff
@@ -0,0 +1,81 @@
+
+  if (showNoDiagnosesScreen) {
+    return <NoDiagnoses />
+  }
```
This doesn't need an effect nor a state:
```javascript
import { isEmpty } from 'lodash'

// ...

if (downloadingDiagnoses) return null // or a loading indicator?
if (isEmpty(diagnoses)) return <NoDiagnoses />

// ...
```

nahudalla commented on src/screens/MainApp/MyDiagnoses/renderUtilities/index.js:
```diff
@@ -0,0 +1,43 @@
+    />
+  )
+}
+
```
I don't like this asynchronous rendering. Rendering is meant to be a synchronous process. In the future, React will introduce concurrent rendering. I don't know how well asynchronous rendering will play with concurrent rendering. If you need to perform asynchronous actions as part of the rendering process of a component, you have to do so when React tells you. That's the purpose of `useEffect`.

MPucheta committed:
Encapsulate ordering logic, introduce lodash utilities, change to synchronous rendering

Async part of the rendering process was moved inside the function in the useEffect hook. Ordering logic is more flexible now, it can accept any SortValue to sort.

MPucheta commented on src/screens/MainApp/MyDiagnoses/index.js:
```diff
@@ -0,0 +1,85 @@
+  return (
+    <View style={styles.container}>
+      <ForceRerenderOnNavigation resetStateFunction={refetchDiagnoses} />
+      {downloadingDiagnoses && <ActivityIndicator size='large' />}
```
I left this indicator here as returning the whole indicator outside would cause a white flicker in the screen (there is a temporal frame between forcing the re-render and the diagnose fetching, so the user could still see the diagnoses for 1 second before having them disappear and appear again, with this here that does not happen and the user gets a better experience.... hopefully)

MPucheta committed:
Fix showing NoDiagnoses on first execution

MPucheta committed:
Remove detailed diagnose files

I do not know what happened, I stashed the files in one branch and they committed in this one

MPucheta committed:
Merge branch 'develop' into MP/MyDiagnoses-Screen
